data/tutorials/language/0lg_08_env.md

```diff
@@ -0,0 +1,557 @@
+
+The scope of a definition is the set of environments where it is reachable.
+
+Although OCaml is an expression-oriented language, it is not entirely free of statements. The `let` construct is a statement. It is a statement which may change the state of the environment by adding a name-value binding. Note that expressions at the toplevel also fall into that category because they are equivalent to `let _ =` definitions.
+```ocaml
+# (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+
+# let _ = (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+```
+
+With respect to the environment, there are no means to:
+- List its contents
+- Remove a definition
+- Reset it to an earlier state
+
+#### Same-Level Shadowing
+
+You can write two definitions using the same identifier in sequence:
+```ocaml
+# let c = 2 * 3;;
+val c : int = 6
+
+# let d = c * 7;;
+val d : int = 42
+
+# let c = 7 * 7;;
+- : int = 49
+
+# d;;
+- : int = 42
+```
+
+In this example, `c` is defined twice. The key thing to understand is `c` is not updated. It looks as if the value of `c` has changed, but it hasn't. When the second definition takes place, the first one becomes unreachable, but it still remains in the environment.
+
+Here is how this can be useful:
+```ocaml
+# let rec length len = function
+  | [] -> len
+  | _ :: u -> length (len + 1) u;;
+val length : int -> 'a list -> int = <fun>
+
+# let length u = length 0 u;;
+val length : 'a list -> int = <fun>
+```
+
+The first definition of `length` has two parameters: the list to process and the accumulated length that has already been computed. Writing it this way makes it tail-recursive.
+
+However, in practice, computing a list's length always starts with the accumulator set to nil. Therefore, the first definition is shadowed by the second where the accumulator parameter is set to 0.
+
+#### Inner Shadowing
+
+A local definition may shadow any other definition, just like a global does. The only difference is that shadowing is limited to the local definition's scope.
+```ocaml
+# let d = 21;;
+val d : int = 21
+
+# let d = 7 in d * 2;;
+val d : int = 14
+
+# d;;
+- : int = 21
+```
+
+## Function Values
+
+### Restrictions 
+
+Functions are supposed to be exactly as other values. However, there are two restrictions.
+
+Function values cannot be displayed in interactive sessions, but the placeholder `<fun>` is displayed. This is because there is nothing meaningful to print. Once parsed and typed-checked, OCaml discards the function's source code and nothing remains to be printed.
+```ocaml
+# sqrt;;
+- : float -> float = <fun>
+```
+
+Equality between functions can't be tested.
+```ocaml
+# pred = succ;;
+Exception: Invalid_argument "compare: functional value".
+```
+
+There are two main reasons explaining this:
+1. It is impossible to write an algorithm that takes two functions and returns `true` if they always return the same output when provided the same input and `false` otherwise
```

Comment:
```suggestion
1. It is impossible to write an algorithm that takes two functions and returns `true`, if they always return the same output when provided the same input, and `false` otherwise
```
Do these commas clarify this sentence or make it less clear? If the latter, we can work on it togherer.